astrolabe/validator.py

```diff
@@ -106,194 +93,395 @@ def run_test_expecting_error(self, driver_workload):
+            if not events['events']:
+                self.fail("The workload executor recorded no events but some were expected")
+
+            numConnectionEvents = 0
+            numPoolEvents = 0
+            numCommandEvents = 0
+
+            for event in events['events']:
+                if ('name' not in event) or (not event['name'].endswith('Event')):
+                    self.fail("The workload executor didn't record event name as expected.")
+
+                if 'observedAt' not in event:
+                    self.fail("The workload executor didn't record event observation time as expected.")
+
+                if event['name'].startswith('Connection'):
+                    numConnectionEvents += 1
+
+                if event['name'].startswith('Pool'):
+                    numPoolEvents += 1
+
+                if event['name'].startswith('Command'):
+                    numCommandEvents += 1
+
+            if numConnectionEvents == 0:
+                self.fail("The workload executor recorded no CMAP connection events but some were expected")
+
+            if numPoolEvents == 0:
+                self.fail("The workload executor recorded no CMAP connection pool events but some were expected")
+
+            if numCommandEvents == 0:
+                self.fail("The workload executor recorded no command monitoring events but some were expected")
+        else:
+            if events['events']:
+                self.fail("The workload executor recorded %d events but none were expected" % len(events['events']))
+
+        # If hasErrorsXorFailures is true, allow either hasErrors or hasFailures
+        # but not both. This is mainly needed for test_num_failures_not_captured
+        # since the workload executor is permitted to report an uncaught failure
+        # as either an error or failure.
+        if hasErrorsXorFailures:
+            # Infer hasErrors and hasFailures from the contents of events.json.
+            # We need only assert xor equivalence here as the function will go
+            # on to assert hasErrors and hasFailures independently.
+            hasErrors = bool(events['errors'])
+            hasFailures = bool(events['failures'])
+
+            if hasErrors and hasFailures:
+                self.fail("The workload executor recorded both errors and failures but only one was expected.")
+
+            if not (hasErrors or hasFailures):
+                self.fail("The workload executor recorded neither errors nor failures but one was expected.")
+
+        # If hasErrors is true, assert that the array is non-empty and that each
+        # object within contains essential fields. If hasErrors is false, the
+        # array should be empty.
+        if hasErrors:
+            if not events['errors']:
+                self.fail("The workload executor recorded no errors but some were expected")
+
+            for error in events['errors']:
+                if ('error' not in error) or ('time' not in error):
```

Comment:
Similar to `observedAt` for events, I chose not to assert the type of `time` here.